---
title: "ChIP-seq of RNAP, sigma factor, RbpA, HelD and CarD in Mycobacterium smegmatis"
author: "Martin Převorovský, prevorov@natur.cuni.cz, http://www.GenoMik.cz"
output: html_document
---

The main aims of this pilot experiment are to:

* determine functionality of RNAP and sigma factor ChIP in STAT phase

* determine and compare genome occupancy of RbpA, HelD and CarD in EXP phase

* determine where along the gene CarD localizes (promoter?)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(GenomicAlignments)
library(GenomicFeatures)

genome_dir <- './genome_Msmegmatis/'
genome_annot <- paste0(genome_dir, 'GCF_000015005.1_ASM1500v1_genomic.customized.gff')
bam_dir <- './BAM_ChIP-seq_Msmegmatis/'
coverage_dir <- './coverage_ChIP-seq_Msmegmatis/'
peak_dir <- './peaks_ChIP-seq/'
image_dir <- './images_Msmegmatis/'
```

Variables for shell scripts are stored in the file "shell_variables_ChIP-seq_Msmegmatis.sh" and can be modified as needed:

```{bash list_variables}
cat shell_variables_ChIP-seq_Msmegmatis.sh
```


# *Mycobacterium smegmatis* MC2 155 genome sequence and annotation

* genome sequence used: gi|118467340|ref|NC_008596.1|

* genome annotation used: genome-build-accession NCBI_Assembly:GCF_000015005.1

* GFF annotation was edited to include the Ms1 ncRNA and the rnpB genes


```{bash, message=FALSE, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

zcat "${genome_dir}GCF_000015005.1_ASM1500v1_genomic.gff.gz" > "${genome_annot}"
sed -i 's/NC_008596.1/gi|118467340|ref|NC_008596.1|/g' "${genome_annot}"
echo -e "gi|118467340|ref|NC_008596.1|\tcustom\tgene\t4392642\t4392993\t.\t-\t.\tID=gene99998;locus_tag=rnpB;gene=rnpB;Name=MSMEG_rnpB" >> "${genome_annot}"
echo -e "gi|118467340|ref|NC_008596.1|\tcustom\tCDS\t4392642\t4392993\t.\t-\t0\tID=cds99998;Parent=gene99998;locus_tag=rnpB" >> "${genome_annot}"
echo -e "gi|118467340|ref|NC_008596.1|\tcustom\tgene\t6242368\t6242672\t.\t+\t.\tID=gene99999;locus_tag=Ms1;gene=Ms1;Name=MSMEG_Ms1" >> "${genome_annot}"
echo -e "gi|118467340|ref|NC_008596.1|\tcustom\tCDS\t6242368\t6242672\t.\t+\t0\tID=cds99999;Parent=gene99999;locus_tag=Ms1" >> "${genome_annot}"

samtools faidx "${genome}"
cut -f1,2 "${genome}.fai" > "${genome_dir}chrom.sizes"
```


# QC of raw sequence reads

* Libraries were sequenced at the Institute of Molecular Genetics, Czech Academy of Sciences, Prague (Illumina NextSeq 550).

* Technical quality of library construction and sequencing run was tested using `fastQC`.

```{bash}
fastqc --version
multiqc --version
```

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

mkdir -p "${QC_dir}"
mapfile -t fastq_files < <(ls -1 "${fastq_dir}" | grep "${fastq_file_ext}")
for i in ${fastq_files[@]};
do
  fastqc --quiet --outdir "${QC_dir}" --threads "${CPU}" "${fastq_dir}${i}"
done

multiqc -f -m fastqc -i ChIP-seq_Msmegmatis -o "${QC_dir}" "${QC_dir}"
```

Results of QC:

* 86 nt SE reads

* 10.3-27.4 M reads per sample

* adapter contamination lower than 0.1%

* all samples give warning on per base sequence content (not sure why)

* 6 samples give warning on per sequence GC content (not sure why)

* all samples give error/warning on high read duplication levels (normal and OK for a small genome)

* read quality scores somewhat deteriorate at the end → 3' end trimming of reads is needed


# Read trimming

* trim parts of the reads where average base quality in a 4 nt window drops below 20

* only keep trimmed reads that are at least 20 nt long

* adapter trimming not needed

```{bash}
java -jar /opt/Trimmomatic-0.39/trimmomatic-0.39.jar -version
```

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

mkdir -p "${fastq_trim_dir}"
mapfile -t fastq_files < <(ls -1 "${fastq_dir}" | grep "${fastq_file_ext}")
for i in ${fastq_files[@]};
do
  java -jar /opt/Trimmomatic-0.39/trimmomatic-0.39.jar SE -phred33 -threads "${CPU}" "${fastq_dir}${i}" "${fastq_trim_dir}${i}.trim.fastq.gz" SLIDINGWINDOW:4:20 MINLEN:20
done
```


# QC of quality-trimmed sequence reads

Effects of trimming were determined by `fastQC`.

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

mkdir -p "${QC_trim_dir}"
mapfile -t fastq_files < <(ls -1 "${fastq_trim_dir}" | grep "${fastq_file_ext}")
for i in ${fastq_files[@]};
do
  fastqc --quiet --outdir "${QC_trim_dir}" --threads "${CPU}" "${fastq_trim_dir}${i}"
done

multiqc -f -m fastqc -i ChIP-seq_Msmegmatis_trimmed -o "${QC_trim_dir}" "${QC_trim_dir}"
```

Results of QC:

* dropped ~2-3% of reads

* 20-86 nt reads (vast majority of reads are >82 nt)

* read quality scores now OK


# Read mapping into *M. smegmatis* genome

```{bash}
hisat2 --version
samtools --version
```

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

# build HISAT2 index
hisat2-build --quiet -p "${CPU}" "${genome}" "${genome}"

# map
mkdir -p "${bam_dir}"
mapfile -t fastq_files < <(ls -1 "${fastq_trim_dir}" | grep "${fastq_file_ext}")
for i in ${fastq_files[@]};
do
	infile="${fastq_trim_dir}${i}"
	outfile="${bam_dir}${i}.bam"
  echo "${outfile}"
	hisat2 -x "${genome}" -U "${infile}" --threads "${CPU}" --no-spliced-alignment -k 10 --summary-file "${outfile}.log" | samtools view -b -F 256 --threads "${CPU}" - | samtools sort --threads "${CPU}" -o "${outfile}" - 
	samtools index -@ "${CPU}" "${outfile}"
	samtools view -c --threads "${CPU}" "${outfile}"
done
```

Notes

* To properly map reads coming from repeat regions (e.g., rRNA), HISAT was set to search for up to 10 matching positions, then only the best (marked as primary by HISAT) alignment was kept.

* Mapping rate ~90-99%.


# Create genome coverage files for each sample 


```{bash}
deeptools --version
```

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

mkdir -p "${coverage_dir}"
mapfile -t bam_files < <(ls -1 "${bam_dir}" | grep "\.bam$")
for i in ${!bam_files[@]};
do
  scaling_factor=`samtools idxstats "${bam_dir}${bam_files[${i}]}" | head -n 1 | cut -f3`
  scaling_factor=`echo "scale=6; 1/(${scaling_factor}/1000000)" | bc`
	input_file="${bam_dir}${bam_files[${i}]}"
	bamCoverage --binSize "${bin_size}" -p "${CPU}" --bam "${input_file}" -o "${coverage_dir}${bam_files[${i}]}.bw" --scaleFactor "${scaling_factor}"
done
```

Notes

* Checked all samples visually in IGV → OK.

* All inputs look similar.

EXP 

* CarD and RbpA profiles look very similar (narrow peaks).

* HelD signal is distributed broadly across a subset of genes, sometimes with a peak-like protrusion at the end (mixed-mode peaks). There is one prominent (tall) broad peak on the HelD locus (MSMEG_2174), suggesting increased copy number of the ORF.

STAT

* RNAP shows many peaks, many of them colocalize with EXP CarD/RbpA.

* Sigma factor profile only partially overlaps with RNAP.


# Peak calling

* Peaks were called for each IP vs the corresponding input sample. This was done individually for each replicate. 

```{bash}
macs2 --version
```

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

gsize=`cat "${genome_dir}chrom.sizes" | cut -f2`
mkdir -p "${peak_dir}"

# STAT phase
macs2 callpeak --treatment "${bam_dir}ChIPseq8_S32_merged_R1_001.fastq.gz.trim.fastq.gz.bam" --control "${bam_dir}ChIPseq7_S31_merged_R1_001.fastq.gz.trim.fastq.gz.bam" --gsize "${gsize}" --outdir "${peak_dir}" --name "STAT_RNAP" --keep-dup auto --nomodel

macs2 callpeak --treatment "${bam_dir}ChIPseq9_S33_merged_R1_001.fastq.gz.trim.fastq.gz.bam" --control "${bam_dir}ChIPseq7_S31_merged_R1_001.fastq.gz.trim.fastq.gz.bam" --gsize "${gsize}" --outdir "${peak_dir}" --name "STAT_sigma" --keep-dup auto --nomodel


# EXP phase
mapfile -t bam_files < <(ls "${bam_dir}"*.bam | grep -v 'ChIPseq[7-9]_' | sort -V)

## CarD
run=0
for i in $(seq 1 6 "${#bam_files[@]}");
do
  run=$((run+1))
  macs2 callpeak --treatment "${bam_files[$i]}" --control "${bam_files[$i-1]}" --gsize "${gsize}" --outdir "${peak_dir}" --name "EXP_CarD_${run}" --keep-dup auto --nomodel
done

## RbpA
run=0
for i in $(seq 3 6 "${#bam_files[@]}");
do
  run=$((run+1))
  macs2 callpeak --treatment "${bam_files[$i]}" --control "${bam_files[$i-1]}" --gsize "${gsize}" --outdir "${peak_dir}" --name "EXP_RbpA_${run}" --keep-dup auto --nomodel
done

## HelD
run=0
for i in $(seq 5 6 "${#bam_files[@]}");
do
  run=$((run+1))
  macs2 callpeak --treatment "${bam_files[$i]}" --control "${bam_files[$i-1]}" --gsize "${gsize}" --outdir "${peak_dir}" --name "EXP_HelD_${run}" --keep-dup auto --nomodel
done
```

* There were not enough peaks to build a peak shift model in many samples, so the `--nomodel` parameter was used throughout (extsize value was left default=200, since the fragment size estimated by `macs2` was around 175-195 anyway (for samples with enough peaks to build the shift model).


# Peak filtering

* Peaks identified in the EXP phase samples were then filtered for those called in all 3 replicates.

* Only the region overlapping between all 3 replicates is kept.

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

# intersect overlapping peaks
bedtools intersect -a "${peak_dir}EXP_CarD_1_peaks.narrowPeak" -b "${peak_dir}EXP_CarD_2_peaks.narrowPeak" | bedtools intersect -a stdin -b "${peak_dir}EXP_CarD_3_peaks.narrowPeak" | bedtools sort -i stdin  > "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak"

bedtools intersect -a "${peak_dir}EXP_RbpA_1_peaks.narrowPeak" -b "${peak_dir}EXP_RbpA_2_peaks.narrowPeak" | bedtools intersect -a stdin -b "${peak_dir}EXP_RbpA_3_peaks.narrowPeak" | bedtools sort -i stdin  > "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak"

bedtools intersect -a "${peak_dir}EXP_HelD_1_peaks.narrowPeak" -b "${peak_dir}EXP_HelD_2_peaks.narrowPeak" | bedtools intersect -a stdin -b "${peak_dir}EXP_HelD_3_peaks.narrowPeak" | bedtools sort -i stdin > "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak"

# merge overlapping peaks (just for info, will not be used for further analyses)
a=`bedtools intersect -wa -a "${peak_dir}EXP_CarD_1_peaks.narrowPeak" -b "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak"`
b=`bedtools intersect -wa -a "${peak_dir}EXP_CarD_2_peaks.narrowPeak" -b "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak"`
c=`bedtools intersect -wa -a "${peak_dir}EXP_CarD_3_peaks.narrowPeak" -b "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak"`
d=( "${a[@]}"$'\n'"${b[@]}"$'\n'"${c[@]}" )
echo "${d[@]}" | bedtools sort -i stdin | bedtools merge -i stdin -c 5,6,9 -o min,distinct,min > "${peak_dir}EXP_CarD_peaks.filtered_merge.bed"
unset a b c d

a=`bedtools intersect -wa -a "${peak_dir}EXP_RbpA_1_peaks.narrowPeak" -b "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak"`
b=`bedtools intersect -wa -a "${peak_dir}EXP_RbpA_2_peaks.narrowPeak" -b "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak"`
c=`bedtools intersect -wa -a "${peak_dir}EXP_RbpA_3_peaks.narrowPeak" -b "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak"`
d=( "${a[@]}"$'\n'"${b[@]}"$'\n'"${c[@]}" )
echo "${d[@]}" | bedtools sort -i stdin | bedtools merge -i stdin -c 5,6,9 -o min,distinct,min > "${peak_dir}EXP_RbpA_peaks.filtered_merge.bed"
unset a b c d

a=`bedtools intersect -wa -a "${peak_dir}EXP_HelD_1_peaks.narrowPeak" -b "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak"`
b=`bedtools intersect -wa -a "${peak_dir}EXP_HelD_2_peaks.narrowPeak" -b "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak"`
c=`bedtools intersect -wa -a "${peak_dir}EXP_HelD_3_peaks.narrowPeak" -b "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak"`
d=( "${a[@]}"$'\n'"${b[@]}"$'\n'"${c[@]}" )
echo "${d[@]}" | bedtools sort -i stdin | bedtools merge -i stdin -c 5,6,9 -o min,distinct,min > "${peak_dir}EXP_HelD_peaks.filtered_merge.bed"
unset a b c d
```

* Adjust q-values and scores in the filtered intersected peaks

As a conservative approach, each final peak is assigned the least significant score/q_value from all replicates.

```{r, eval=TRUE}
extraCols_narrowPeak <- c(signal_value = "numeric", p_value = "numeric",
                          q_value = "numeric", summit_offset = "integer")

# CarD
peaks.intersect <- import.bed(paste0(peak_dir, 'EXP_CarD_peaks.filtered_intersect.narrowPeak'),
                              extraCols = extraCols_narrowPeak)
peaks_1 <- import.bed(paste0(peak_dir, 'EXP_CarD_1_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_2 <- import.bed(paste0(peak_dir, 'EXP_CarD_2_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_3 <- import.bed(paste0(peak_dir, 'EXP_CarD_3_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks <- c(peaks_1, peaks_2, peaks_3)
for (i in 1:length(peaks.intersect)){
  peaks_overlap.tmp <- peaks[to(findOverlaps(peaks.intersect[i], peaks))]
  peaks.intersect[i]$score <- min(peaks_overlap.tmp$score, na.rm = TRUE)
  peaks.intersect[i]$q_value <- min(peaks_overlap.tmp$q_value, na.rm = TRUE)
}
write.table(data.frame(seqname = seqnames(peaks.intersect),
                       start = start(peaks.intersect) - 1,
                       end = end(peaks.intersect), 
                       name = peaks.intersect$name,
                       score = peaks.intersect$score,
                       strand = '.',
                       q_value = peaks.intersect$q_value),
            paste0(peak_dir, 'EXP_CarD_peaks.filtered_intersect.p-value.bed'),
            col.names = FALSE, row.names = FALSE, quote = FALSE, sep = '\t')
rm(peaks.intersect, peaks_1, peaks_2, peaks_3, peaks, peaks_overlap.tmp)

# HelD
peaks.intersect <- import.bed(paste0(peak_dir, 'EXP_HelD_peaks.filtered_intersect.narrowPeak'),
                              extraCols = extraCols_narrowPeak)
peaks_1 <- import.bed(paste0(peak_dir, 'EXP_HelD_1_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_2 <- import.bed(paste0(peak_dir, 'EXP_HelD_2_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_3 <- import.bed(paste0(peak_dir, 'EXP_HelD_3_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks <- c(peaks_1, peaks_2, peaks_3)
for (i in 1:length(peaks.intersect)){
  peaks_overlap.tmp <- peaks[to(findOverlaps(peaks.intersect[i], peaks))]
  peaks.intersect[i]$score <- min(peaks_overlap.tmp$score, na.rm = TRUE)
  peaks.intersect[i]$q_value <- min(peaks_overlap.tmp$q_value, na.rm = TRUE)
}
write.table(data.frame(seqname = seqnames(peaks.intersect),
                       start = start(peaks.intersect) - 1,
                       end = end(peaks.intersect), 
                       name = peaks.intersect$name,
                       score = peaks.intersect$score,
                       strand = '.',
                       q_value = peaks.intersect$q_value),
            paste0(peak_dir, 'EXP_HelD_peaks.filtered_intersect.p-value.bed'),
            col.names = FALSE, row.names = FALSE, quote = FALSE, sep = '\t')
rm(peaks.intersect, peaks_1, peaks_2, peaks_3, peaks, peaks_overlap.tmp)

# RbpA
peaks.intersect <- import.bed(paste0(peak_dir, 'EXP_RbpA_peaks.filtered_intersect.narrowPeak'),
                              extraCols = extraCols_narrowPeak)
peaks_1 <- import.bed(paste0(peak_dir, 'EXP_RbpA_1_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_2 <- import.bed(paste0(peak_dir, 'EXP_RbpA_2_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks_3 <- import.bed(paste0(peak_dir, 'EXP_RbpA_3_peaks.narrowPeak'), extraCols = extraCols_narrowPeak)
peaks <- c(peaks_1, peaks_2, peaks_3)
for (i in 1:length(peaks.intersect)){
  peaks_overlap.tmp <- peaks[to(findOverlaps(peaks.intersect[i], peaks))]
  peaks.intersect[i]$score <- min(peaks_overlap.tmp$score, na.rm = TRUE)
  peaks.intersect[i]$q_value <- min(peaks_overlap.tmp$q_value, na.rm = TRUE)
}
write.table(data.frame(seqname = seqnames(peaks.intersect),
                       start = start(peaks.intersect) - 1,
                       end = end(peaks.intersect), 
                       name = peaks.intersect$name,
                       score = peaks.intersect$score,
                       strand = '.',
                       q_value = peaks.intersect$q_value),
            paste0(peak_dir, 'EXP_RbpA_peaks.filtered_intersect.p-value.bed'),
            col.names = FALSE, row.names = FALSE, quote = FALSE, sep = '\t')
rm(peaks.intersect, peaks_1, peaks_2, peaks_3, peaks, peaks_overlap.tmp)
```


# Peak analysis

* Create a BED file of all ORFs for plotting average-gene coverage profiles.

* Create a GFF file for TSS for identification of genes closest to peaks.

```{r, eval=TRUE}
gff <- import.gff3(genome_annot)
gff.CDS <- gff[gff$type == 'CDS', ]
gff.CDS$score <- 0
export.bed(gff.CDS, paste0(genome_dir, 'all_ORFs.bed'))
median(width(gff.CDS))

gff.TSS <- gff[gff$type == 'gene', ] # this dataset is created to preserve locus_tags in the gff subselection
gff.TSS <- gff.TSS[gff.TSS$ID %in% unlist(gff.CDS$Parent), ]
end(gff.TSS[strand(gff.TSS) == '+', ]) <- start(gff.TSS[strand(gff.TSS) == '+', ]) + 1
start(gff.TSS[strand(gff.TSS) == '-', ]) <- end(gff.TSS[strand(gff.TSS) == '-', ]) - 1
export.gff3(gff.TSS[order(start(gff.TSS)), ], paste0(genome_dir, 'all_TSS.gff3'))
```

* Determine overlaps between target sites of the studied proteins.

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

wc -l "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak" "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak" "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak"
bedtools intersect -u -a "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak" -b "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak" | wc -l
bedtools intersect -u -b "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak" -a "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak" | wc -l

bedtools intersect -u -a "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak" -b "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak" | wc -l
bedtools intersect -u -b "${peak_dir}EXP_CarD_peaks.filtered_intersect.narrowPeak" -a "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak" | wc -l

bedtools intersect -u -a "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak" -b "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak" | wc -l
bedtools intersect -u -b "${peak_dir}EXP_RbpA_peaks.filtered_intersect.narrowPeak" -a "${peak_dir}EXP_HelD_peaks.filtered_intersect.narrowPeak" | wc -l
```

* Create average-gene coverage profiles.

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

mkdir -p "${image_dir}"

# whole genes
computeMatrix scale-regions -S "${coverage_dir}ChIPseq2_S26_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq11_S35_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq17_S41_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq4_S28_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq13_S37_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq19_S43_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq6_S30_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq15_S39_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq21_S45_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" -R "${genome_dir}all_ORFs.bed" --beforeRegionStartLength 1000 --regionBodyLength 850 --afterRegionStartLength 1000 --outFileName "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs.mat.gz" --numberOfProcessors "${CPU}"

plotProfile -m "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs.mat.gz" -out "${image_dir}plotProfile_ChIP-seq.all_ORFs.svg" --regionsLabel "all ORFs" --samplesLabel "CarD_1" "CarD_2" "CarD_3" "RbpA_1" "RbpA_2" "RbpA_3" "HelD_1" "HelD_2" "HelD_3" --numPlotsPerRow 3

plotProfile -m "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs.mat.gz" -out "${image_dir}plotProfile_ChIP-seq.all_ORFs2.svg" --regionsLabel "all ORFs" --samplesLabel "CarD_1" "CarD_2" "CarD_3" "RbpA_1" "RbpA_2" "RbpA_3" "HelD_1" "HelD_2" "HelD_3" --perGroup

# TSS
computeMatrix reference-point -S "${coverage_dir}ChIPseq2_S26_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq11_S35_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq17_S41_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq4_S28_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq13_S37_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq19_S43_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq6_S30_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq15_S39_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq21_S45_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" -R "${genome_dir}all_ORFs.bed" --referencePoint TSS --beforeRegionStartLength 400 --afterRegionStartLength 200 --outFileName "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs_TSS.mat.gz" --numberOfProcessors "${CPU}"

plotProfile -m "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs_TSS.mat.gz" -out "${image_dir}plotProfile_ChIP-seq.all_ORFs_TSS.svg" --regionsLabel "all ORFs" --samplesLabel "CarD_1" "CarD_2" "CarD_3" "RbpA_1" "RbpA_2" "RbpA_3" "HelD_1" "HelD_2" "HelD_3" --perGroup

computeMatrix reference-point -S "${coverage_dir}ChIPseq2_S26_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq11_S35_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq17_S41_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq4_S28_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq13_S37_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" "${coverage_dir}ChIPseq19_S43_merged_R1_001.fastq.gz.trim.fastq.gz.bam.bw" -R "${genome_dir}all_ORFs.bed" --referencePoint TSS --beforeRegionStartLength 400 --afterRegionStartLength 200 --outFileName "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs_TSS2.mat.gz" --numberOfProcessors "${CPU}"

plotProfile -m "${coverage_dir}computeMatrix_ChIP-seq.all_ORFs_TSS2.mat.gz" -out "${image_dir}plotProfile_ChIP-seq.all_ORFs_TSS2.svg" --regionsLabel "all ORFs" --samplesLabel "CarD_1" "CarD_2" "CarD_3" "RbpA_1" "RbpA_2" "RbpA_3" --perGroup
```

* Find closest downstream ORFs

```{bash, eval=TRUE}
source shell_variables_ChIP-seq_Msmegmatis.sh

bedtools closest -t first -a "${peak_dir}EXP_CarD_peaks.filtered_intersect.p-value.bed" -b "${genome_dir}all_TSS.gff3" > "${peak_dir}EXP_CarD_peaks_CLOSEST.bed"
bedtools closest -t first -a "${peak_dir}EXP_RbpA_peaks.filtered_intersect.p-value.bed" -b "${genome_dir}all_TSS.gff3" > "${peak_dir}EXP_RbpA_peaks_CLOSEST.bed"
bedtools closest -t first -a "${peak_dir}EXP_HelD_peaks.filtered_intersect.p-value.bed" -b "${genome_dir}all_TSS.gff3" > "${peak_dir}EXP_HelD_peaks_CLOSEST.bed"
```

* Integrate ChIP-seq peak data with RNA-seq expression data (ArrayExpress accession E-MTAB-7004 ; 3 replicates of WT_EXP: ERR2696389, ERR2696390, ERR2696391)

```{r, eval=TRUE}
rnaseq <- read.csv('RNA-seq_Ms1_counts_q-filtered_run3.norm.txt', row.names = 1)
rnaseq <- rowMeans(rnaseq[, c('WT_exp_1', 'WT_exp_2', 'WT_exp_3')])
rnaseq[rnaseq == 0] <- NA
rnaseq <- log2(rnaseq)

peaks.RbpA <- read.delim(paste0(peak_dir, 'EXP_RbpA_peaks_CLOSEST.bed'), header = FALSE, stringsAsFactors = FALSE)
peaks.RbpA$locus_tag <- substr(sub('.*locus_tag=', '', peaks.RbpA[, 19]), 1, 10)
peaks.CarD <- read.delim(paste0(peak_dir, 'EXP_CarD_peaks_CLOSEST.bed'), header = FALSE, stringsAsFactors = FALSE)
peaks.CarD$locus_tag <- substr(sub('.*locus_tag=', '', peaks.CarD[, 19]), 1, 10)
peaks.HelD <- read.delim(paste0(peak_dir, 'EXP_HelD_peaks_CLOSEST.bed'), header = FALSE, stringsAsFactors = FALSE)
peaks.HelD$locus_tag <- substr(sub('.*locus_tag=', '', peaks.HelD[, 19]), 1, 10)

pdf(paste0(image_dir, 'comparisons_with_RNA-seq_WT_EXP.pdf'))

stripchart(rnaseq[!names(rnaseq) %in% peaks.RbpA$locus_tag], method = 'jitter', pch = 20, col = rgb(0, 0, 0, 0.2), vertical = TRUE, 
           at = 0.7, group.names = 'RbpA')
stripchart(rnaseq[!names(rnaseq) %in% peaks.CarD$locus_tag], method = 'jitter', pch = 20, col = rgb(0, 0, 0, 0.2), vertical = TRUE, 
           add = TRUE, at = 1, group.names = 'CarD')
stripchart(rnaseq[!names(rnaseq) %in% peaks.HelD$locus_tag], method = 'jitter', pch = 20, col = rgb(0, 0, 0, 0.2), vertical = TRUE, 
           add = TRUE, at = 1.3, group.names = 'HelD')
stripchart(rnaseq[names(rnaseq) %in% peaks.RbpA$locus_tag], 
           method = 'jitter', pch = 20, col = rgb(1, 0, 0, 0.2), add = TRUE, vertical = TRUE, at = 0.7)
stripchart(rnaseq[names(rnaseq) %in% peaks.CarD$locus_tag], 
           method = 'jitter', pch = 20, col = rgb(1, 0, 0, 0.2), add = TRUE, vertical = TRUE, at = 1)
stripchart(rnaseq[names(rnaseq) %in% peaks.HelD$locus_tag], 
           method = 'jitter', pch = 20, col = rgb(1, 0, 0, 0.2), add = TRUE, vertical = TRUE, at = 1.3)
mtext('RbpA___CarD___HelD')

boxplot(rnaseq, rnaseq, rnaseq, names = c('RbpA', 'CarD', 'HelD'), las = 3)
stripchart(rnaseq[names(rnaseq) %in% peaks.RbpA$locus_tag], col = rgb(0, 0, 1, 0.2), add = TRUE, vertical = TRUE, method = 'jitter', pch = 20, at = 1)
stripchart(rnaseq[names(rnaseq) %in% peaks.CarD$locus_tag], col = rgb(0, 0, 1, 0.2), add = TRUE, vertical = TRUE, method = 'jitter', pch = 20, at = 2)
stripchart(rnaseq[names(rnaseq) %in% peaks.HelD$locus_tag], col = rgb(0, 0, 1, 0.2), add = TRUE, vertical = TRUE, method = 'jitter', pch = 20, at = 3)

for (i in 1:nrow(peaks.RbpA)){
  peaks.RbpA[i, 'rnaseq'] <- rnaseq[names(rnaseq) == peaks.RbpA[i, 'locus_tag']]
}
for (i in 1:nrow(peaks.CarD)){
  peaks.CarD[i, 'rnaseq'] <- rnaseq[names(rnaseq) == peaks.CarD[i, 'locus_tag']]
}
for (i in 1:nrow(peaks.HelD)){
  peaks.HelD[i, 'rnaseq'] <- rnaseq[names(rnaseq) == peaks.HelD[i, 'locus_tag']]
}

plot(log10(peaks.RbpA[, 9]), peaks.RbpA$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'log10(peak q-value)', ylab = 'log2(norm_expression)', main = 'RbpA peaks and their closest genes')
abline(lm(peaks.RbpA$rnaseq ~ log10(peaks.RbpA[, 9])))
cor.test(log10(peaks.RbpA[, 9]), peaks.RbpA$rnaseq, use = 'complete')
plot(log10(peaks.CarD[, 9]), peaks.CarD$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'log10(peak q-value)', ylab = 'log2(norm_expression)', main = 'CarD peaks and their closest genes')
abline(lm(peaks.CarD$rnaseq ~ log10(peaks.CarD[, 9])))
cor.test(log10(peaks.CarD[, 9]), peaks.CarD$rnaseq, use = 'complete')
plot(log10(peaks.HelD[, 9]), peaks.HelD$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'log10(peak q-value)', ylab = 'log2(norm_expression)', main = 'HelD peaks and their closest genes')
abline(lm(peaks.HelD$rnaseq ~ log10(peaks.HelD[, 9])))
cor.test(log10(peaks.HelD[, 9]), peaks.HelD$rnaseq, use = 'complete')

r <- as.data.frame(rnaseq)
r.RbpA <- as.data.frame(r[rownames(r) %in% peaks.RbpA$locus_tag, ])
colnames(r.RbpA) <- 'rnaseq'
r.CarD <- as.data.frame(r[rownames(r) %in% peaks.CarD$locus_tag, ])
colnames(r.CarD) <- 'rnaseq'
r.HelD <- as.data.frame(r[rownames(r) %in% peaks.HelD$locus_tag, ])
colnames(r.HelD) <- 'rnaseq'

library(ggplot2)
library(patchwork)
p1 <- ggplot(r, aes(x = 1, y = rnaseq)) +
  geom_violin() +
  labs(x = 'RbpA') + 
  geom_jitter(data = r.RbpA, alpha = 0.2, aes(x = 1, y = rnaseq), color = 'blue', width = 0.2)
p2 <- ggplot(r, aes(x = 1, y = rnaseq)) +
  geom_violin() +
  labs(x = 'CarD') + 
  geom_jitter(data = r.CarD, alpha = 0.2, aes(x = 1, y = rnaseq), color = 'blue', width = 0.2)
p3 <- ggplot(r, aes(x = 1, y = rnaseq)) +
  geom_violin() +
  labs(x = 'HelD') + 
  geom_jitter(data = r.HelD, alpha = 0.2, aes(x = 1, y = rnaseq), color = 'blue', width = 0.2)
p1 + p2 + p3


plot(peaks.RbpA[, 7], peaks.RbpA$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'coverage enrichment at peak', 
     ylab = 'log2(norm_expression)', main = 'RbpA peaks and their closest genes')
abline(lm(peaks.RbpA$rnaseq ~ peaks.RbpA[, 7]))
cor.test(peaks.RbpA[, 7], peaks.RbpA$rnaseq, use = 'complete')
plot(peaks.CarD[, 7], peaks.CarD$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'coverage enrichment at peak', 
     ylab = 'log2(norm_expression)', main = 'CarD peaks and their closest genes')
abline(lm(peaks.CarD$rnaseq ~ peaks.CarD[, 7]))
cor.test(peaks.CarD[, 7], peaks.CarD$rnaseq, use = 'complete')
plot(peaks.HelD[, 7], peaks.HelD$rnaseq, 
     pch = 20, col = rgb(0, 0, 0, 0.4), xlab = 'coverage enrichment at peak', 
     ylab = 'log2(norm_expression)', main = 'HelD peaks and their closest genes')
abline(lm(peaks.HelD$rnaseq ~ peaks.HelD[, 7]))
cor.test(peaks.HelD[, 7], peaks.HelD$rnaseq, use = 'complete')

dev.off()
```


***

```{r}
sessionInfo()
```
